Remove workaround in Access.defineClass()

keithc-ca opened the pull request:
* Add RAM class flag J9ClassIsExemptFromValidation which is set for MagicAccessorImpl and its subclasses which makes them exempt for some access and validation rules (like classes defined via unsafe API).
* Add and use macro: J9CLASS_IS_EXEMPT_FROM_VALIDATION.
* New flag, J9_FINDCLASS_FLAG_NAME_IS_INVALID, is used to signal an invalid class name if the actual loaded class is not exempt. If the class being defined is not exempt, defineClassCommon() will now return NULL without a pending exception. ClassLoader_defineClassImpl() throws an exception only in that case.
* Allow JavaLangAccess to define a class with a null ProtectionDomain.
* Map proxy classes to the proper module.
* defineClassCommon() now calls fixUnsafeMethods() as necessary
* fix loop start index in verifyQualifiedName()

Fixes: #3641
Fixes: #4661

gacholio commented on runtime/oti/j9javaaccessflags.h:
```diff
@@ -67,7 +67,7 @@
 #define J9AccClassReferenceWeak 0x10000000
 #define J9AccClassRomToRamMask 0xF3000000
-#define J9AccClassUnsafe 0x40000
+#define J9AccClassUnsafe 0x40000 /* obsolete - do not use */
```
Typcially we define J9AccClassUnused0x4000 or something like that. If you haven't deleted the constant, you can't prove you found all the references.

keithc-ca replied:
It's still referenced in DDR code for reading older core files.

gacholio commented on runtime/vm/KeyHashTable.c:
```diff
@@ -447,7 +446,7 @@ hashPkgTableIDFor(J9VMThread *vmThread, J9ClassLoader *classLoader, J9ROMClass *
 	key.tag = (UDATA)romClass | TAG_ROM_CLASS;
 
-	if (isSystemClassLoader && (J9ROMCLASS_IS_UNSAFE(romClass) || (LOAD_LOCATION_UNKNOWN == locationType))) {
+	if (isSystemClassLoader && (LOAD_LOCATION_UNKNOWN == locationType)) {
```
This change is not equivalent code.

keithc-ca replied:
I realize that; it's not clear what better we can do given the 'unsafe' bit is no longer available (and wasn't entirely reliable).

keithc-ca replied:
I've restored setting the 'unsafe' bit via that path to the tag bit can continue to be set in the same situations.

gacholio commented on runtime/vm/createramclass.cpp:
```diff
@@ -3010,6 +3028,18 @@ internalCreateRAMClassFromROMClass(J9VMThread *vmThread, J9ClassLoader *classLoa
+			const U_16 magicLength = sizeof(MAGIC_ACCESSOR_IMPL) - 1; /* exclude NUL */
+			J9UTF8 *className = J9ROMCLASS_CLASSNAME(romClass);
+
+			if ((magicLength == J9UTF8_LENGTH(className)) &&
```
We have macros for this - `J9UTF8_LITERAL_EQUALS(data1, length1, cString)`

gacholio commented:
Let's make `(J9_ARE_NO_BITS_SET(resultClass->classFlags, J9ClassIsExemptFromValidation)` into a macro, which would make many of your unnecessary formatting changes even more unnecessary.

keithc-ca commented:
I added `J9CLASS_IS_EXEMPT_FROM_VALIDATION` in `j9.h` and used it as suggested.
I also made use of `J9UTF8_LITERAL_EQUALS`.

DanHeidinga commented on runtime/jcl/common/clsldr.cpp:
```diff
@@ -30,36 +30,42 @@
 Java_java_lang_ClassLoader_defineClassImpl(JNIEnv *env, jobject receiver, jstring className, jbyteArray classRep, jint offset, jint length, jobject protectionDomain)
 {
+	J9VMThread *currentThread = (J9VMThread *)env;
+	jclass result = defineClassCommon(env, receiver, className, classRep, offset, length, protectionDomain, 0, NULL);
```
Why did you refactor this code to define the class before doing the verification of the name?  Once the class has been defined, it gets added to classloader tables and the like - I would expect other threads to be able to find it as though it was successfully defined even though it might fail below now.

keithc-ca replied:
Whether the class is exempt from the usual name rules is dependent on whether it extends `MagicAccessorImpl`: that information isn't available before the call to `
defineClassCommon`. I think you're right, though, if the name is bad, it ought to be removed from those tables.

keithc-ca replied:
Those class cannot be found because `Class.forName()` also makes use of `verifyQualifiedName()`.

DanHeidinga replied:
Need to check the resolveClassRef path as well

keithc-ca replied:
I've restored validation of the name before the call, but deferred throwing the exception until we know the class is not exempt.

DanHeidinga commented on runtime/vm/createramclass.cpp:
```diff
@@ -1519,11 +1529,11 @@ popFromClassLoadingStack(J9VMThread *vmThread)
 {
 	J9JavaVM *vm = vmThread->javaVM;
-	const BOOLEAN isROMClassUnsafe = (J9ROMCLASS_IS_UNSAFE(romClass) != 0);
+	BOOLEAN isExemptFromValidation = J9_ARE_ANY_BITS_SET(options, J9_FINDCLASS_FLAG_UNSAFE);
```
why was const removed?

keithc-ca replied:
It will be set below if the superclass is exempt.

DanHeidinga commented on runtime/vm/createramclass.cpp:
```diff
@@ -1571,22 +1580,28 @@ loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9
-				if ((!superClassIsPublic && (packageID != superclass->packageID))
-					|| (superClassIsPublic && (J9_VISIBILITY_ALLOWED != checkModuleAccess(vmThread, vm, romClass, module, superclass->romClass, superclass->module, superclass->packageID, 0)))
+				if (superClassIsPublic
+						? (J9_VISIBILITY_ALLOWED != checkModuleAccess(vmThread, vm, romClass, module, superclass->romClass, superclass->module, superclass->packageID, 0))
```
Please revert this refactoring - embedding a ternary if inside an if statement is painful from a readability perspective.  Repeating the single boolean check is fine if the compiler isn't smart enough to remove it

keithc-ca replied:
I disagree: I find it clearer and easier to understand this way. I will revert if you insist.

keithc-ca replied:
I will revert this change (and the similar one below): we can return to this discussion later.

DanHeidinga commented on runtime/vm/createramclass.cpp:
```diff
@@ -1614,18 +1629,19 @@ loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9
-						if ((!interfaceIsPublic && (packageID != interfaceClass->packageID))
-							|| (interfaceIsPublic && (J9_VISIBILITY_ALLOWED != checkModuleAccess(vmThread, vm, romClass, module, interfaceClass->romClass, interfaceClass->module, interfaceClass->packageID, 0)))
+						if (interfaceIsPublic
+								? (J9_VISIBILITY_ALLOWED != checkModuleAccess(vmThread, vm, romClass, module, interfaceClass->romClass, interfaceClass->module, interfaceClass->packageID, 0))
```
Same here - embedding ternary ifs in if statements is a readability concern - please revert this refactoring

keithc-ca commented:
I believe I've addressed all the comments (so far) in a new branch (extending this one) and will push to this branch once I've verified it works as expected.

keithc-ca commented:
Ready for another round of reviews.

DanHeidinga commented:
Jenkins test sanity plinux,win jdk11

DanHeidinga commented:
Jenkins test extended zlinux jdk8

DanHeidinga commented:
Jenkins test sanity xlinux jdk12

DanHeidinga commented:
Jenkins test extended zlinux jdk8

jdekonin commented:
jenkins test sanity win jdk11

jdekonin commented:
jenkins test sanity xlinux jdk12

keithc-ca commented:
I'd like a chance to squash this before it's merged, please.

pshipton commented:
Technically the 0.13 PR should "fix" the listed Issues.

DanHeidinga commented:
Running the builds first as I'm reasonably certain this looks right but wanted a broad set of tests first.

@keithc-ca Will ping you to squash to before merge

DanHeidinga commented:
@keithc-ca I'm seeing some jvmti redefinition-related test failures in the extended jdk8 build.  Can you take a look?

keithc-ca commented:
I am reworking the changes to `clsldr.cpp` and `createramclass.cpp` to restore the exception behavior as expected by the failing test (and to match the RI).

keithc-ca commented:
I have the right exception behavior; now looking at the JVMTI redefinition issue.

gacholio commented:
Another thing I thought of over the weekend - classes which extend a class defined via Unsafe do not currently inherit the validation exclusion. This case is unlikely, but I think it's definitely possible.

keithc-ca commented:
The redefinition issue should now be fixed.

keithc-ca commented:
Jenkins test extended zlinux jdk8

gacholio commented:
jenkins test sanity xlinux jdk8

keithc-ca commented:
I've reworked this so classes which extend those defined via unsafe API aren't exempted from the normal validation rules. After running local testing, I'll push to this branch.

gacholio commented:
Sanity build failed because master is unstable. Will try again later.

gacholio commented:
OK, I'll try testing again after review.

gacholio commented:
May be worth rebasing on master after your latest changes to avoid build issues.

keithc-ca commented:
Testing went well, so I squashed and rebased. Hopefully this is in its final state.

pshipton commented:
jenkins test sanity zlinux jdk8,jdk12

pshipton commented:
jenkins test sanity plinux jdk11

pshipton commented:
jenkins test extended plinux jdk8,jdk11,jdk12

pshipton commented:
jenkins test sanity win jdk8

pshipton commented:
The jdk12 sanity testing is failing.

gacholio commented on runtime/vm/createramclass.cpp:
```diff
@@ -1584,15 +1594,26 @@ loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9
 				}
 			}
 
+			if (J9_ARE_ANY_BITS_SET(options, J9_FINDCLASS_FLAG_NAME_IS_INVALID) && !isExemptFromValidation) {
```
Should this be setting the bit and letting the caller deal with the exception? The exception below is a much better one than is being thrown in the new calling code.

keithc-ca replied:
When you say 'below', do you mean the one in `clsldr.cpp`? Just returning `FALSE` here with no exception seems to work (and is reasonable given the caller has claimed responsibility for dealing with the invalid name).

gacholio replied:
"below" I mean the special exception setter call which has good detail info. The one being thrown outside is bare (no message). I'd move the good call out to the caller and check there (which you indicate works).

keithc-ca replied:
I spoke too soon: calling `ClassLoader.loadClass()` with a `/` name referring to an already loaded class yields a `LinkageError` instead of the more specific `NoClassDefFoundError`. I'll see if there's a straight-forward fix for that.

gacholio replied:
Strictly speaking, that's fine - you are throwing a LinkageError. Unless there's a test querying the exact class of the exception, I think that should be fine.

Comment welcome from @DanHeidinga @pshipton

DanHeidinga replied:
While GAC is likely technically correct, the javadoc for loadClass() only mentions `ClassNotFoundException` so I expect there's a bunch of code that knows to map `NoClassDefFoundError` -> `ClassNotFoundException` and may not handle LinkageError correctly

keithc-ca replied:
OpenJ9 has at least one test that expects `NoClassDefFoundError` and considers the more general `LinkageError` a failure: that expected behavior is consistent with the RI.

keithc-ca replied:
I only know of the reverse mapping (`ClassNotFoundException` -> `NoClassDefFoundError`) that used to be involved in the compiled code for expressions of the form `Whatever.class` before class constants were a thing in class files: the compiler knew about the mentioned class and so generates code to catch `ClassNotFoundException` (a checked exception) thrown by `Class.forName()` and map it to `NoClassDefFoundError`.

keithc-ca commented:
This version passes all sanity testing in both jdk11 and jdk12.

pshipton commented:
jenkins test extended plinux jdk11,jdk12

gacholio commented:
jenkins test all xlinux jdk8

gacholio commented:
jenkins test sanity zlinux jdk11,jdk12

gacholio commented:
jenkins compile win jdk12

gacholio commented on sourcetools/j9constantpool/com/ibm/oti/VMCPTool/ConstantPoolStream.java:
```diff
@@ -329,7 +329,7 @@ private void writeHeader(int size) {
 
 			String shapeDescriptionOffset = "sizeof(J9ROMClass) - offsetof(J9ROMClass, cpShapeDescription) + " + size;
-			out.println("\t{0,0,0,0,0,J9AccClassUnsafe, 0,0,0,0,0,0,0,0, " + itemCount + ", " + itemCount + ", 0,0,0, " + shapeDescriptionOffset + ", 0,},");
+			out.println("\t{0,0,0,0,0,0, 0,0,0,0,0,0,0,0, " + itemCount + ", " + itemCount + ", 0,0,0, " + shapeDescriptionOffset + ", 0,},");
```
This change is unnecessary now (since we're keeping unsafe).

keithc-ca replied:
Reverted.

gacholio commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -2879,7 +2880,6 @@ typedef struct J9ROMImageHeader {
 
 /* @ddr_namespace: map_to_type=J9ClassLocation */
 
-struct J9Class;
```
Unnecessary.

keithc-ca replied:
Reverted.

gacholio commented on runtime/jcl/common/stdinit.c:
```diff
@@ -111,6 +111,9 @@ standardInit( J9JavaVM *vm, char *dllName)
 
+	memset(&jclFakeClass, 0, sizeof(jclFakeClass));
+	/* the fake class must be exempt because it references package-private classes */
+	jclFakeClass.classFlags = J9ClassIsExemptFromValidation;
```
This should also be unnecessary since we're not removing unsafe.

keithc-ca replied:
Reverted.

gacholio commented:
The comments in the header are out-of-date (not removing unsafe).

I'd like @DanHeidinga to have a look at this before merge.

keithc-ca commented:
Description updated.

pshipton commented:
For the record, all the testing started today has passed.

DanHeidinga commented on runtime/vm/createramclass.cpp:
```diff
@@ -2934,21 +2957,41 @@ internalCreateRAMClassFromROMClass(J9VMThread *vmThread, J9ClassLoader *classLoa
+					if (classLoader != javaVM->systemClassLoader) {
+						findModule = true;
+					} else if ((LOAD_LOCATION_PATCH_PATH == locationType)
+							|| (LOAD_LOCATION_MODULE == locationType)) {
```
Formatting:
```c
} else if ((LOAD_LOCATION_PATCH_PATH == locationType)
|| (LOAD_LOCATION_MODULE == locationType)
) {
        ........
}
```

keithc-ca replied:
Fixed.

DanHeidinga commented on runtime/vm/createramclass.cpp:
```diff
@@ -2934,21 +2957,41 @@ internalCreateRAMClassFromROMClass(J9VMThread *vmThread, J9ClassLoader *classLoa
+							/*
+							 * Proxy classes loaded by the system classloader may belong to a dynamically
+							 * created module (which has been granted access to the interfaces it must implement).
+							 */
```
is this a bug fix for something beyond the Access.defineClass issue?  If so, please add it to the issue description

DanHeidinga commented:
overall, this looks good to me.  Was there something you specific you wanted me to look at @gacholio ?

gacholio commented:
> overall, this looks good to me. Was there something you specific you wanted me to look at @gacholio ?

Nope, just wanted to make sure you agreed with the overall strategy.

gacholio commented on jcl/src/java.base/share/classes/java/lang/ClassLoader.java:
```diff
@@ -537,6 +537,18 @@ private String checkClassName(String className) {
 		ProtectionDomain protectionDomain) 
 		throws java.lang.ClassFormatError 
+{
+	return defineClassInternal(className, classRep, offset, length, protectionDomain, false);
```
Extreme nitpick - code above uses a variable for the boolean, this code does not.

gacholio replied:
I'd personally prefer an inline value with a comment to minimize bytecode size when running in the interpreter.

keithc-ca replied:
Comments added at both call sites.

gacholio commented:
Some local test failures are being investigated before this can be merged.

keithc-ca commented:
None of those local test failures are introduced by this PR.
